chatserver/src/session.rs

```diff
@@ -0,0 +1,117 @@
+/// instances via `HttpContext::state()`
+pub struct WsChatSessionState {
+    pub addr: Addr<Syn, server::ChatServer>,
+    pub log: Addr<Syn, logging::MozLogger>,
+}
+
+pub struct WsChatSession {
+    /// unique session id
+    pub id: usize,
+    /// Client must send ping at least once per 10 seconds, otherwise we drop
+    /// connection.
+    pub hb: Instant,
+    /// joined channel
+    pub channel: Uuid,
+    /// peer name
+    pub name: Option<String>,
+}
+
+impl Actor for WsChatSession {
+    type Context = ws::WebsocketContext<Self, WsChatSessionState>;
+
+    /// Method is called on actor start.
+    /// We register ws session with ChatServer
+    fn started(&mut self, ctx: &mut Self::Context) {
+        // register self in chat server. `AsyncContext::wait` register
+        // future within context, but context waits until this future resolves
+        // before processing any other events.
+        // HttpContext::state() is instance of WsChatSessionState, state is shared
+        // across all routes within application
+        let addr: Addr<Syn, _> = ctx.address();
+        ctx.state()
+            .addr
+            .send(server::Connect {
+                addr: addr.recipient(),
+                channel: self.channel.clone(),
+            })
+            .into_actor(self)
+            .then(|res, act, ctx| {
+                match res {
+                    Ok(res) => act.id = res,
+                    // something is wrong with chat server
+                    Err(err) => {
+                        ctx.state().log.do_send(logging::LogMessage {
+                            level: logging::ErrorLevel::Error,
+                            msg: format!("{:?}", err),
+                        });
+                        ctx.stop()
+                    }
+                }
+                fut::ok(())
+            })
+            .wait(ctx);
+    }
+
+    fn stopping(&mut self, ctx: &mut Self::Context) -> Running {
+        // notify chat server
+        ctx.state().addr.do_send(server::Disconnect {
+            channel: self.channel.clone(),
+            id: self.id,
+        });
+        Running::Stop
+    }
+}
+
+/// Handle messages from chat server, we simply send it to peer websocket
+impl Handler<server::TextMessage> for WsChatSession {
+    type Result = ();
+
+    fn handle(&mut self, msg: server::TextMessage, ctx: &mut Self::Context) {
+        ctx.text(msg.0);
+    }
+}
+
+/// WebSocket message handler
+impl StreamHandler<ws::Message, ws::ProtocolError> for WsChatSession {
+    fn handle(&mut self, msg: ws::Message, ctx: &mut Self::Context) {
+        ctx.state().log.do_send(logging::LogMessage {
+            level: logging::ErrorLevel::Info,
+            msg: format!("Websocket Message: {:?}", msg),
+        });
+        match msg {
+            ws::Message::Ping(msg) => ctx.pong(&msg),
+            ws::Message::Pong(msg) => self.hb = Instant::now(),
```

Comment:
Could probably leave this out, or null op the pong. The example code that used self.hb left this in but removed the important func that actually does the heartbeat check. Though if you do want a proper heart-beat.... here's the PR and location of code that makes it work:
https://github.com/actix/examples/commit/16f7649b29d2d7ad3a8acdf00ccf2678030e509d#diff-faf4a6b3e58832518cf0e42565c9635cR148

Reply:
I'll admit that I didn't really pay a lot of attention to this when I copied it over from the example. For real production, you're right, we want this to be more solid. I'll add a note to indicate that it's half complete.